# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/arm/Lowering-arm.cpp
# Commit: 83a3ed07220d
# Full Hash: 83a3ed07220dfe4c8a2d16a2c48cfa8d49635362
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-14 09:51:01
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.3: Establish dependency from tls for arm callWithABI div/mod i64. r=lth
#   
#   To be able to call c++ runtime via Builtin thunk we need to set up WasmTlsReg.
#   In this patch I create dependencies from MIR level to Codegen to be sure that WasmTlsReg is alive
#   when we call runtime in div/mod i64 for arm.
# ==============================================================================

diff -r ca66708049b8 -r 83a3ed07220d js/src/jit/arm/Lowering-arm.cpp
--- a/js/src/jit/arm/Lowering-arm.cpp	Mon Sep 14 02:54:53 2020 +0000
+++ b/js/src/jit/arm/Lowering-arm.cpp	Mon Sep 14 04:00:18 2020 +0000
@@ -409,45 +409,51 @@
 }
 
 void LIRGeneratorARM::lowerDivI64(MDiv* div) {
+  MOZ_CRASH("We use MWasmBuiltinDivI64 instead.");
+}
+
+void LIRGeneratorARM::lowerWasmBuiltinDivI64(MWasmBuiltinDivI64* div) {
   if (div->isUnsigned()) {
-    lowerUDivI64(div);
+    LUDivOrModI64* lir =
+        new (alloc()) LUDivOrModI64(useInt64RegisterAtStart(div->lhs()),
+                                    useInt64RegisterAtStart(div->rhs()),
+                                    useFixedAtStart(div->tls(), WasmTlsReg));
+    defineReturn(lir, div);
     return;
   }
 
   LDivOrModI64* lir = new (alloc()) LDivOrModI64(
-      useInt64RegisterAtStart(div->lhs()), useInt64RegisterAtStart(div->rhs()));
+      useInt64RegisterAtStart(div->lhs()), useInt64RegisterAtStart(div->rhs()),
+      useFixedAtStart(div->tls(), WasmTlsReg));
   defineReturn(lir, div);
 }
 
-void LIRGeneratorARM::lowerWasmBuiltinDivI64(MWasmBuiltinDivI64* div) {
-  MOZ_CRASH("We don't use runtime div for this architecture");
+void LIRGeneratorARM::lowerModI64(MMod* mod) {
+  MOZ_CRASH("We use MWasmBuiltinModI64 instead.");
 }
 
-void LIRGeneratorARM::lowerModI64(MMod* mod) {
+void LIRGeneratorARM::lowerWasmBuiltinModI64(MWasmBuiltinModI64* mod) {
   if (mod->isUnsigned()) {
-    lowerUModI64(mod);
+    LUDivOrModI64* lir =
+        new (alloc()) LUDivOrModI64(useInt64RegisterAtStart(mod->lhs()),
+                                    useInt64RegisterAtStart(mod->rhs()),
+                                    useFixedAtStart(mod->tls(), WasmTlsReg));
+    defineReturn(lir, mod);
     return;
   }
 
   LDivOrModI64* lir = new (alloc()) LDivOrModI64(
-      useInt64RegisterAtStart(mod->lhs()), useInt64RegisterAtStart(mod->rhs()));
+      useInt64RegisterAtStart(mod->lhs()), useInt64RegisterAtStart(mod->rhs()),
+      useFixedAtStart(mod->tls(), WasmTlsReg));
   defineReturn(lir, mod);
 }
 
-void LIRGeneratorARM::lowerWasmBuiltinModI64(MWasmBuiltinModI64* mod) {
-  MOZ_CRASH("We don't use runtime mod for this architecture");
-}
-
 void LIRGeneratorARM::lowerUDivI64(MDiv* div) {
-  LUDivOrModI64* lir = new (alloc()) LUDivOrModI64(
-      useInt64RegisterAtStart(div->lhs()), useInt64RegisterAtStart(div->rhs()));
-  defineReturn(lir, div);
+  MOZ_CRASH("We use MWasmBuiltinDivI64 instead.");
 }
 
 void LIRGeneratorARM::lowerUModI64(MMod* mod) {
-  LUDivOrModI64* lir = new (alloc()) LUDivOrModI64(
-      useInt64RegisterAtStart(mod->lhs()), useInt64RegisterAtStart(mod->rhs()));
-  defineReturn(lir, mod);
+  MOZ_CRASH("We use MWasmBuiltinModI64 instead.");
 }
 
 void LIRGenerator::visitPowHalf(MPowHalf* ins) {